Coupon Management: Networking - Remote with loadAllCoupons endpoint method

joshheald-a8c opened the pull request:
Part of #3912
Please ensure #3987 is merged before this PR

## Description
This PR adds `CouponsRemote`, with a single endpoint method for the time being: `loadAllCoupons`, which makes `GET` requests to `/coupons/` and supports pagination parameters. This method will required by the coupon list screens, to list the coupons available for a site.

The implementation here follows the existing pattern, but some further improvement to allow injection of the mapper to the remote would be helpful. This would let us test for what the siteID passed to the mapper is independently of the results of parsing coupons. I did attempt to improve on this while working on the code, but it turned out to have a bit too much complexity to do in the scope of this work. I'd be interested to improve on this in future though.

## Testing
The endpoint is not in use yet, so no applicable testing other than the unit tests, and CI once it's merged to the main repo.

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

joshheald-a8c committed:
CouponsRemote with loadAllCoupons method

The Remote is closely coupled with the Mapper here, which I’d like to improve for testability.

The implementation here follows the existing pattern, but some further improvement to allow injection of the mapper would be helpful, so that we could test for what the siteID passed to the mapper is independently of the results of parsing coupons.

CouponsRemote with loadAllCoupons method

The Remote is closely coupled with the Mapper here, which I’d like to improve for testability.

The implementation here follows the existing pattern, but some further improvement to allow injection of the mapper would be helpful, so that we could test for what the siteID passed to the mapper is independently of the results of parsing coupons.

joshheald-a8c committed:
CouponsRemote tested for NetworkError passing

joshheald-a8c committed:
Use siteID rather than siteId

pmusolino commented:
@joshheald-a8c can I ask you to solve the conflicts before the review, please? 🙇

joshheald-a8c committed:
Change loadAllCoupons to use Result<> enqueue call

Found that the Result<[Coupon], Error> approach is preferred, so this change converts the existing request to use that method.

joshheald-a8c committed:
Update CouponRemote tests to use `waitFor` syntax

`waitFor` seems to give a nicer flow to the tests because of removing one optional, and some of the boilerplate around expectations.

joshheald-a8c committed:
CouponsRemote with loadAllCoupons method

The Remote is closely coupled with the Mapper here, which I’d like to improve for testability.

The implementation here follows the existing pattern, but some further improvement to allow injection of the mapper would be helpful, so that we could test for what the siteID passed to the mapper is independently of the results of parsing coupons.

joshheald-a8c commented:
> @joshheald-a8c can I ask you to solve the conflicts before the review, please? 🙇

@pmusolino ready to go now. I'll look at the following ones and try to sort them out before you get to them too.

pmusolino commented on Networking/NetworkingTests/Remote/CouponsRemoteTests.swift:
```diff
@@ -0,0 +1,136 @@
+import XCTest
+@testable import Networking
+import Alamofire
```
nit: the import of Alamofire seems not needed here.

pmusolino commented on Networking/NetworkingTests/Remote/CouponsRemoteTests.swift:
```diff
@@ -0,0 +1,136 @@
+@testable import Networking
+import Alamofire
+
+class CouponsRemoteTests: XCTestCase {
```
```suggestion
final class CouponsRemoteTests: XCTestCase {
```

pmusolino reviewed (approved):
LGTM, just two small nits 💯

joshheald-a8c committed:
CouponsRemoteTests improvements from code review

joshheald-a8c commented:
@pmusolino Thanks, sorted those 😊
